protocol/kafka_confluent/v2/protocol.go

```diff
@@ -0,0 +1,225 @@
+	}
+	defer in.Finish(err)
+
+	kafkaMsg := &kafka.Message{
+		TopicPartition: kafka.TopicPartition{
+			Topic:     &p.producerDefaultTopic,
+			Partition: kafka.PartitionAny,
+		},
+	}
+
+	if topic := cecontext.TopicFrom(ctx); topic != "" {
+		kafkaMsg.TopicPartition.Topic = &topic
+	}
+
+	if messageKey := MessageKeyFrom(ctx); messageKey != "" {
+		kafkaMsg.Key = []byte(messageKey)
+	}
+
+	err = WriteProducerMessage(ctx, in, kafkaMsg, transformers...)
+	if err != nil {
+		return err
+	}
+
+	err = p.producer.Produce(kafkaMsg, p.producerDeliveryChan)
+	if err != nil {
+		return err
+	}
+	e := <-p.producerDeliveryChan
+	m := e.(*kafka.Message)
+	if m.TopicPartition.Error != nil {
+		return m.TopicPartition.Error
+	}
+	return nil
+}
+
+func (p *Protocol) OpenInbound(ctx context.Context) error {
+	if p.consumer == nil {
+		return fmt.Errorf("the consumer client must be set")
+	}
+	if p.consumerTopics == nil {
+		return fmt.Errorf("the consumer topics must be set")
+	}
+
+	p.consumerMux.Lock()
+	defer p.consumerMux.Unlock()
+	logger := cecontext.LoggerFrom(ctx)
+
+	// Query committed offsets for each partition
+	if positions := TopicPartitionOffsetsFrom(ctx); positions != nil {
+		if err := p.consumer.Assign(positions); err != nil {
+			return err
+		}
+	}
+
+	logger.Infof("Subscribing to topics: %v", p.consumerTopics)
+	err := p.consumer.SubscribeTopics(p.consumerTopics, p.consumerRebalanceCb)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if !p.consumer.IsClosed() {
+			logger.Infof("Closing consumer %v", p.consumerTopics)
+			if err = p.consumer.Close(); err != nil {
+				logger.Errorf("failed to close the consumer: %v", err)
+			}
+		}
+	}()
+
+	for !p.consumer.IsClosed() {
+		select {
+		case <-ctx.Done():
+			logger.Debugf("The consumer context is canceled!")
+			return nil
+		default:
+			ev := p.consumer.Poll(p.consumerPollTimeout)
+			if ev == nil {
+				continue
+			}
+			switch e := ev.(type) {
+			case *kafka.Message:
+				p.incoming <- e
+			case kafka.Error:
+				// Errors should generally be considered informational, the client will try to automatically recover.
+				// But in here, we choose to terminate the application if all brokers are down.
+				logger.Infof("Error %v: %v", e.Code(), e)
+				if p.consumerErrorHandler != nil {
+					p.consumerErrorHandler(e)
+				}
+				if e.Code() == kafka.ErrAllBrokersDown {
+					logger.Debugf("All broker connections are down!")
+					return e
+				}
+			default:
+				logger.Debugf("Ignored the kafka.Event: %v", e)
+			}
+		}
+	}
+	return fmt.Errorf("the consumer is closed")
+}
+
+// Receive implements Receiver.Receive
+func (p *Protocol) Receive(ctx context.Context) (binding.Message, error) {
+	select {
+	case m, ok := <-p.incoming:
+		if !ok {
+			return nil, io.EOF
+		}
+		msg := NewMessage(m)
+		return msg, nil
+	case <-ctx.Done():
+		return nil, io.EOF
+	}
+}
+
+func (p *Protocol) Close(ctx context.Context) error {
+	if p.consumer != nil && !p.consumer.IsClosed() {
```

Comment:
Now you have a data race, if this is called from another goroutine. You need to protect with mutex here as well and be careful with deadlocking

Reply:
I added another lock(mutex) to handle the data race between the `Opener` and `Closer`. It will make sure that the `p.consumer.Close()` and `p.consumer().IsClose()` will don't execute in parallel.